File: ruby/easy/k_diff_pairs_in_an_array.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def find_pairs(nums, k)
    nums.sort!
    i, count = 0, 0
    while i < nums.length
        num = nums[i]
        i += 1
        if binary_search(i, nums.length-1, nums, (num + k))
            count += 1
        end
        while i < nums.length && num == nums[i]
           i += 1
        end
    end
    return count
end

def binary_search(start, stop, nums, target)
    while start <= stop
        middle = (start + stop)/2
        if(nums[middle] == target)
            return true
        elsif nums[middle] > target
            stop = middle-1
        else
            start = middle+1
        end
    end
    return false
end

File: javascript/easy/kDiffPairsInAnArray.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findPairs = function(nums, k) {
    if(nums.length === 0) return 0;
    
    nums.sort(function(x,y){ return x-y; });
    
    let i = 0, num = nums[0], count = 0;
    if(binarySearch(i+1, nums.length-1, nums, (num+k))) {
        count++;   
    }
    
    while(i < nums.length) {
        if(nums[i] === num) {
            i++; continue;
        }
        num = nums[i];
        if(binarySearch(i+1, nums.length-1, nums, (num+k))) {
            count++;   
        }
        i++;
    }
    return count;
};

var binarySearch = function(start, end, nums, target) {
    while(start <= end) {
        let middle = parseInt((start + end)/2);
        if(nums[middle] === target) {
            return true;
        } else if(nums[middle] > target) {
            end = middle-1;
        } else {
            start = middle+1;
        }
    }
    return false;
};
